src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperUsageDescriptorFactory.cs

```diff
@@ -0,0 +1,171 @@
+                    CultureInfo.CurrentCulture,
+                    assemblyDirectory,
+                    assemblyXmlDocumentationName);
+                FileInfo assemblyXmlDocumentationFile;
+
+                if (localizedXmlFile == null)
+                {
+                    // Check for an uncultured XML file next to the assembly
+                    assemblyXmlDocumentationFile = new FileInfo(
+                        Path.Combine(assemblyDirectory, assemblyXmlDocumentationName));
+                }
+                else
+                {
+                    assemblyXmlDocumentationFile = localizedXmlFile;
+                }
+
+                return assemblyXmlDocumentationFile.Exists == true ? assemblyXmlDocumentationFile : null;
+            }
+            catch (ArgumentException)
+            {
+                // Could not resolve XML file.
+                return null;
+            }
+        }
+
+        private static IEnumerable<string> ExpandCulture(CultureInfo culture)
+        {
+            // Following the fall-back process defined by:
+            // https://msdn.microsoft.com/en-us/library/sb6a8618.aspx#cpconpackagingdeployingresourcesanchor1
+            do
+            {
+                yield return culture.Name;
+
+                culture = culture.Parent;
+            } while (culture != null && culture != CultureInfo.InvariantCulture);
+        }
+
+        private static FileInfo GetLocalizedXmlDocumentationFile(
+            CultureInfo culture,
+            string assemblyDirectory,
+            string assemblyXmlDocumentationName)
+        {
+            var cultureNames = ExpandCulture(culture);
+            var localizedXmlDocumentationFile = cultureNames
+                .Select(cultureName =>
+                {
+                    var cultureSpecificFileName =
+                    Path.ChangeExtension(assemblyXmlDocumentationName, cultureName + ".xml");
+
+                    // Look for a culture specific XML file next to the assembly.
+                    var xmlFile = new FileInfo(Path.Combine(assemblyDirectory, cultureSpecificFileName));
+
+                    if (!xmlFile.Exists)
+                    {
+                        // Look for an XML file with the same name as the assembly in a culture specific directory.
+                        xmlFile = new FileInfo(
+                            Path.Combine(assemblyDirectory, cultureName, assemblyXmlDocumentationName));
+
+                        if (!xmlFile.Exists)
+                        {
+                            // Look for a culture specific XML file in a culture specific directory.
+                            xmlFile = new FileInfo(
+                                Path.Combine(assemblyDirectory, cultureName, cultureSpecificFileName));
+                        }
+                    }
+
+                    return xmlFile;
+                }).FirstOrDefault(file => file.Exists);
```

Comment:
Looks like an unrolled loop a compiler has spit out in IL. Would be easier to understood if the `ExpandPaths()` method expanded all the paths and the Linq was just `Select(path => new FileInfo(path)).FirstOrDefault(file => file.Exists)`. Also avoids the extra `Exists` check and makes the linked page more relevant within `ExpandPaths()`. The method might be
``` c#
private static IEnumerable<string> ExpandPaths(
    CultureInfo culture,
    string assemblyDirectory,
    string assemblyXmlDocumentationName)
{
    do
    {
        var cultureName = culture.Name;
        var cultureSpecificFileName =
            Path.ChangeExtension(assemblyXmlDocumentationName, cultureName + ".xml");

        yield return Path.Combine(assemblyDirectory, cultureSpecificFileName);
        yield return Path.Combine(assemblyDirectory, cultureName, assemblyXmlDocumentationName));
        yield return Path.Combine(assemblyDirectory, cultureName, cultureSpecificFileName));

        culture = culture.Parent;
    } while (culture != null && culture != CultureInfo.InvariantCulture);
}
```